src/WebJobs.Script.WebHost/Metrics/FlexConsumptionMetricsPublisher.cs

```diff
@@ -0,0 +1,301 @@
+        {
+            _standbyOptions = standbyOptions ?? throw new ArgumentNullException(nameof(standbyOptions));
+            _environment = environment ?? throw new ArgumentNullException(nameof(environment));
+            _optionsMonitor = optionsMonitor ?? throw new ArgumentNullException(nameof(optionsMonitor));
+            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+            _fileSystem = fileSystem ?? new FileSystem();
+
+            if (_standbyOptions.CurrentValue.InStandbyMode)
+            {
+                _logger.LogDebug("Registering StandbyOptions change subscription.");
+                _standbyOptionsOnChangeSubscription = _standbyOptions.OnChange(o => OnStandbyOptionsChange());
+            }
+            else
+            {
+                Start();
+            }
+        }
+
+        // internal properties for testing
+        internal long FunctionExecutionCount { get; set; }
+
+        internal long FunctionExecutionTimeMS { get; set; }
+
+        internal long ActiveFunctionCount { get; set; }
+
+        internal bool IsAlwaysReady { get; set; }
+
+        internal string MetricsFilePath { get; set; }
+
+        public void Start()
+        {
+            Initialize();
+
+            _logger.LogInformation($"Starting metrics publisher (AlwaysReady={IsAlwaysReady}, MetricsPath='{MetricsFilePath}').");
+
+            _metricsPublisherTimer = new Timer(OnFunctionMetricsPublishTimer, null, _initialPublishDelay, _metricPublishInterval);
+            _started = true;
+        }
+
+        /// <summary>
+        /// Initialize any environmentally derived state after specialization, prior to starting the publisher.
+        /// </summary>
+        internal void Initialize()
+        {
+            _metricPublishInterval = TimeSpan.FromMilliseconds(_optionsMonitor.CurrentValue.MetricsPublishIntervalMS);
+            _initialPublishDelay = TimeSpan.FromMilliseconds(_optionsMonitor.CurrentValue.InitialPublishDelayMS);
+            _intervalStopwatch = ValueStopwatch.StartNew();
+            MetricsFilePath = _optionsMonitor.CurrentValue.MetricsFilePath;
+
+            IsAlwaysReady = _environment.GetEnvironmentVariable(EnvironmentSettingNames.FunctionsAlwaysReadyInstance) == "1";
+        }
+
+        internal async Task OnPublishMetrics()
+        {
+            try
+            {
+                if (FunctionExecutionCount == 0 && FunctionExecutionTimeMS == 0 && !IsAlwaysReady)
+                {
+                    // no activity to report
+                    return;
+                }
+
+                // we've been accumulating function activity for the entire period
+                // publish this activity and reset
+                Metrics metrics = null;
+                lock (_lock)
+                {
+                    metrics = new Metrics
+                    {
+                        TotalTimeMS = (long)_intervalStopwatch.GetElapsedTime().TotalMilliseconds,
```

Comment:
curious why we track this interval, is it check for any gaps in receiver side?

basically we just need to know in every 5 second interval
-how long was atleast 1 function active
right?

Reply:
oh i see. is it for always ready?

Reply:
Yes, it's for AlwaysReady :) And generally it's just being precise about the duration of the metering interval. While I could have hard coded it to be the timer interval, I wanted it to be more precise in the face of timer slowdowns, etc.